File: Python/OOP/Card-Game.py
from CSNEA_Shared import *
import random


class Card:
    def __init__(self, colour: str, number: int):
        self.colour = colour
        self.number = number

    def __repr__(self):
        return self.colour + " " + str(self.number)


def win_hand(player_name: str, card_list: list):
    """Append cards to winner's stack and report win."""
    # Appends both cards to the winner's card stack
    card_list.append(player1.active_card)
    card_list.append(p2ActiveCard)
    print(f"{player_name} won that hand!")
    print()
    input("Press enter to continue")
    print()


colourDict = {
    'Red,Black' or 'Black,Red': 'Red',
    'Yellow,Red' or 'Red,Yellow': 'Yellow',
    'Black,Yellow' or 'Yellow,Black': 'Black'
}


def compare(card1: Card, card2: Card):
    """Compare cards to find winner of hand."""
    if card1.colour == card2.colour:
        if card1.number > card2.number:
            win_hand(player1.name, p1Cards)
        else:
            win_hand(player2.name, p2Cards)

        return

    colour = card1.colour + "," + card2.colour
    win_colour = colourDict.get(colour)
    if win_colour == card1.colour:
        win_hand(player1.name, p1Cards)
    else:
        win_hand(player2.name, p2Cards)


# ===== Welcome & Rules

print("Welcome to The Card Game!")
print("In this game, each player draws a card, the cards are compared and the winner takes both cards.")
print("In total, 15 hands are drawn.")
print("The player with the most cards at the end wins.")
print()
addPlayerFlag = input("Press 1 to add a new player. Press enter to log in. ")
if addPlayerFlag == 1:
    add_player()
print()

# Authenticate players
player1, player2 = authenticate_two_players()

# ===== The Game

# Initialises player's card stacks as empty
p1Cards = []
p2Cards = []

input("Press enter to start")
print("Let's begin the game!")
print()

# Create list of all cards as Card objects
with open("deck.csv") as f:
    deckList = [Card(line.split(",")[0], int(line.split(",")[1])) for line in f.read().splitlines()]

random.shuffle(deckList)

handNum = 1

while len(deckList) > 0:
    print(f"Hand: {handNum}")
    player1.active_card = deckList[0]
    del deckList[0]
    p2ActiveCard = deckList[0]
    del deckList[0]

    print(f"{player1.name} drew a {player1.active_card}")
    print(f"{player2.name} drew a {p2ActiveCard}")
    print()
    input("Press enter to continue")
    print()

    compare(player1.active_card, p2ActiveCard)

    handNum = handNum + 1

print("All cards have been drawn!")
input("The winner is...")
print()

if len(p1Cards) > len(p2Cards):
    winner = player1.name
    winNum = len(p1Cards)
    win_cards = p1Cards
else:
    winner = player2.name
    winNum = len(p2Cards)
    win_cards = p2Cards

print(f"{winner}! With {winNum} cards!")
print()
input("They had these cards:")
print()

for card in win_cards:
    print(card)

print()

write_score("card_game_scores.csv", winner + "," + str(winNum))

display_high_scores("card_game_scores.csv")

print()
input("Press enter to finish")
print()
print(f"Thank you, {player1.name} and {player2.name}, for playing The Card Game!")
input("Goodbye!")

File: Java/CSNEA_Shared/src/main/org/dyson/games/CardGame.java
package org.dyson.games;

import java.io.*;
import java.util.*;

record Card(String colour, int number) {

	public String getColour() { return colour; }

	public int getNumber() { return number; }

	public String displayName() { return this.colour + " " + this.number; }
}

public class CardGame {

	static Player player1;
	static Player player2;

	static Card p1ActiveCard;
	static Card p2ActiveCard;

	// Create ArrayLists for players' card stacks
	static ArrayList<Card> p1Cards = new ArrayList<>();
	static ArrayList<Card> p2Cards = new ArrayList<>();

	// Create scanner object to read input
	static Scanner inputScanner = new Scanner(System.in);

	public static void winHand(String name, ArrayList<Card> cardStack) {
		// Add active Cards to player's card stack
		cardStack.add(p1ActiveCard);
		cardStack.add(p2ActiveCard);

		System.out.println(name + " won that hand!");
		System.out.println();
		System.out.println("Press enter to continue.");
		inputScanner.nextLine();
	}

	public static void compare(Card card1, Card card2) {
		// If colours are equal, higher number wins
		if (card1.getColour().equals(card2.getColour())) {
			// If the colours are equal, then the numbers must be different, so we don't have to worry about a bias
			if (card1.getNumber() > card2.getNumber()) {
				winHand(player1.getName(), p1Cards);
			} else {
				winHand(player2.getName(), p2Cards);
			}
		} else {
			// If numbers are the same, compare colours to find winner
			String colour = card1.getColour() + " " + card2.getColour();
			String winColour = "";

			if (colour.equals("Red Black") | colour.equals("Black Red")) {
				winColour = "Red";
			} else if (colour.equals("Yellow Red") | colour.equals("Red Yellow")) {
				winColour = "Yellow";
			} else if (colour.equals("Black Yellow") | colour.equals("Yellow Black")) {
				winColour = "Black";
			}

			// Use winColour to find winning player
			if (winColour.equals(card1.getColour())) {
				winHand(player1.getName(), p1Cards);
			} else if (winColour.equals(card2.getColour())) {
				winHand(player2.getName(), p2Cards);
			} else {
				// If neither player wins, print error message
				System.out.println("Unhandled winColour exception in CardGame.compare().");
			}
		}
	}

	public static void main(String[] args) {
		System.out.println("Welcome to The Card Game!");
		System.out.println("In this game, each player draws a card, the cards are compared and the winner takes both cards.");
		System.out.println();
		System.out.println("Press 1 to add a new player. Press enter to log in.");

		// Call addPlayer() if input == "1"
		if (inputScanner.hasNextLine()) {
			if (inputScanner.nextLine().equals("1")) {
				GameMethods.addPlayer();
			}
		}

		player1 = GameMethods.authenticateReturnPlayer(1);
		player2 = GameMethods.authenticateReturnPlayer(2);

		// If same account, quit program
		if (player1.getName().equals(player2.getName())) {
			System.out.println("Sorry, " + player2.getName() + ", but that's the same account as player 1.");
			System.out.println("Press enter to exit");
			inputScanner.nextLine();
			System.exit(0);
		}

		// Create ArrayList of Cards to allow for shuffling later
		ArrayList<Card> deckList = new ArrayList<>();

		// Read deck.csv into deckList
		try {
			File deckFile = new File("deck.csv");
			Scanner deckScanner = new Scanner(deckFile);

			while (deckScanner.hasNextLine()) {
				String line = deckScanner.nextLine();
				deckList.add(new Card(line.split(",")[0], Integer.parseInt(line.split(",")[1])));
			}
		}
		catch (FileNotFoundException e) {
			System.out.println("deck.csv not found.");
			e.printStackTrace();
		}

		// Shuffle deckList and create normal array called deck
		Collections.shuffle(deckList);
		Card[] deck = deckList.toArray(new Card[0]);

		System.out.println("Press enter to start.");
		inputScanner.nextLine();
		System.out.println("Let's begin the game!");
		inputScanner.nextLine();

		int handNum = 1;

		// Loop over deck and compare top 2 Cards every time
		for (int i = 0; i < deck.length; i += 2) {
			System.out.println("Hand: " + handNum);
			p1ActiveCard = deck[i];
			p2ActiveCard = deck[i + 1];

			System.out.println(player1.getName() + " drew a " + p1ActiveCard.displayName());
			System.out.println(player2.getName() + " drew a " + p2ActiveCard.displayName());
			System.out.println();
			System.out.println("Press enter to continue.");
			inputScanner.nextLine();

			// Compare Cards to find winner
			compare(p1ActiveCard, p2ActiveCard);

			handNum++;
		}

		System.out.println("All cards have been drawn!");
		System.out.println("The winner is...");
		inputScanner.nextLine();

		// Create variables for winner
		Player winner;
		int winNum;
		ArrayList<Card> winCards;

		// Find winner from length of card stacks
		if (p1Cards.size() > p2Cards.size()) {
			winner = player1;
			winNum = p1Cards.size();
			winCards = p1Cards;
		} else {
			winner = player2;
			winNum = p2Cards.size();
			winCards = p2Cards;
		}

		winner.setScore(winNum); // Set the winner's score value to allow it to be written to a file properly

		System.out.println(winner.getName() + "! With " + winNum + " cards!");
		System.out.println();
		System.out.println("They had these cards:");
		inputScanner.nextLine();

		// Print all Cards
		for (Card card : winCards) {
			System.out.println(card.displayName());
		}

		System.out.println();

		GameMethods.writeScore("card_game_scores.csv", winner);

		GameMethods.displayHighScores("card_game_scores.csv");

		System.out.println();
		System.out.println("Press enter to finish.");
		inputScanner.nextLine();
		System.out.println("Thank you, " + player1.getName() + " and " + player2.getName() + ", for playing The Card Game!");
		inputScanner.nextLine();
	}

}
